super-ddf: optimize DDF header search for widely used RAID controllers

slayercat opened the pull request:
**Added logic to check the last 512 bytes of the device (`dsize -
  512`) first**
for the presence of the DDF header (`DDF_HEADER_MAGIC`).
  - If the header is found, the function immediately proceeds without performing a broader search.
  - Validates the magic value after reading the 512 bytes.

**Fallback to searching the last 32MB of the device**

 using the  `search_for_ddf_headers` function when the last 512 bytes do not contain the header.
  - Performs a block-by-block search to locate the DDF header within the specified region.


see also: https://github.com/md-raid-utilities/mdadm/issues/135

mtkaczyk commented:
HI @slayercat 
Contribution here requires understanding of `kernel coding style`.
There are a lot of comments made by our automation, please address them first (make review job passing).

mtkaczyk commented on super-ddf.c:
```diff
@@ -877,30 +877,151 @@ static void *load_section(int fd, struct ddf_super *super, void *buf,
+	const size_t READ_BLOCK_SIZE = 4096;
+	const unsigned long long SEARCH_REGION_SIZE = 32 * 1024 * 1024;
+
+	// Determine the search range
```
For comments, you should use one of:
```
/*Body*/

/*
 * Body
 */
```

ncroxon replied:
@slayercat 
Please check https://github.com/md-raid-utilities/mdadm/pull/136/checks

ncroxon commented:
@slayercat 
Please check https://github.com/md-raid-utilities/mdadm/pull/136/checks for errors...

slayercat commented:
Thank you for pointing out the issues with the code style. I have reviewed and modified the submitted code snippets.

mtkaczyk commented:
```
super-ddf: optimize DDF header search for widely used RAID controllers
**Added logic to check the last 512 bytes of the device (`dsize -
  512`) first**
for the presence of the DDF header (`DDF_HEADER_MAGIC`).
  - If the header is found, the function immediately proceeds without
    performing a broader search.
  - Validates the magic value after reading the 512 bytes.

**Fallback to searching the last 32MB of the device**
using the  `search_for_ddf_headers` function when the last 512 bytes
do not contain the header.
  - Performs a block-by-block search to locate the DDF header
    within the specified region.

See also: https://github.com/md-raid-utilities/mdadm/issues/135

Signed-off-by: lilinzhe <llz@antiy.cn>
```

Kernel style is not only about the code style. You should also follow commit rules, described here:
https://www.kernel.org/doc/html/v4.17/process/submitting-patches.html#describe-your-changes

I think that the description is good but you should divide it into 2 patches. Please do not use markup in commit body, you can keep it in PR description because I'm always cherry-picking commits and PR description is irrelevant.

Sorry, for that! I'm not trying to be nitpicker. As long as mailing list contribution is welcomed we trying to keep similar level of commits quality!

If you are interested, You can explore list online archive to find see comments from kernel folks on mdadm patches (you can search by "mdadm"):
https://lore.kernel.org/linux-raid/

I hope, this exercise will be useful for you to be better prepared for linux kernel contributions :)

mtkaczyk commented on super-ddf.c:
```diff
@@ -877,30 +877,170 @@ static void *load_section(int fd, struct ddf_super *super, void *buf,
+
+cleanup:
+	free(buffer);		/* Free allocated memory */
+	return result;		/* Return the result */
```
Comment is not needed.

mtkaczyk reviewed (changes requested):
Hello,
I did a quick code review. Please see my comments.

Thanks,
Mariusz

slayercat commented on super-ddf.c:
```diff
@@ -877,30 +877,170 @@ static void *load_section(int fd, struct ddf_super *super, void *buf,
+
+	/* Allocate aligned memory for reading */
+	if (posix_memalign(&buffer, READ_BLOCK_SIZE, READ_BLOCK_SIZE) != 0) {
+		perror("posix_memalign");
```
Thank you for identifying that. I have submitted a new function named `xmemalign` in `xmalloc.c to handle posix_memalign, which exits if the allocation fails.

However, there are many direct function calls in the `ddf` code, so it will take some time to transition to using this new function.

slayercat commented on super-ddf.c:
```diff
@@ -877,30 +877,170 @@ static void *load_section(int fd, struct ddf_super *super, void *buf,
-	if (lseek64(fd, dsize - 512, 0) == -1L) {
-		if (devname)
-			pr_err("Cannot seek to anchor block on %s: %s\n",
+	const size_t READ_BLOCK_SIZE = 4096;
```
Thank you for addressing that. I have moved all variable declarations to the beginning of the function and replaced the const declarations with macros to align with the existing code practices.

slayercat commented on super-ddf.c:
```diff
@@ -877,30 +877,170 @@ static void *load_section(int fd, struct ddf_super *super, void *buf,
 
-static int load_ddf_headers(int fd, struct ddf_super *super, char *devname)
+
+/* Search for DDF_HEADER_MAGIC in the last 32MB of the device */
```
Thank you for addressing that. I have added the reason to the comments at the beginning of search_for_ddf_headers.

```
/*
 * Search for DDF_HEADER_MAGIC in the last 32MB of the device
 *
 * According to the specification, the Anchor Header should be placed at
 * the end of the disk. However,some widely used RAID hardware, such as
 * LSI and PERC, do not position it within the last 512 bytes of the disk.
 *
 */
static int search_for_ddf_headers(int fd, char *devname,
				  unsigned long long *out)
```

slayercat commented on super-ddf.c:
```diff
@@ -5195,6 +5195,19 @@ static void default_geometry_ddf(struct supertype *st, int *level, int *layout,
+			 int uuid_set, char *homehost)
+{
+	/*
+	 * A dummy update_super function is needed
```
Thank you for addressing that. I have added the reason in the commit message and found a related discussion about this in the maillist.


```
commit e3d493d7046cc07b4eca95a2cc24cc59e0ee3b34
Author: lilinzhe <llz@antiy.cn>
Date:   Mon Dec 16 12:00:02 2024 +0800

    super-ddf: Prevent crash when handling DDF metadata
    
    A dummy function is defined because availability of ss->update_super is
    not always verified.
    
    This fix addresses a crash reported when assembling a RAID array using
    mdadm with DDF metadata. For more details, see the discussion at:
    https://lore.kernel.org/all/
    CALHdMH30LuxR4tz9jP2ykDaDJtZ3P7L3LrZ+9e4Fq=Q6NwSM=Q@mail.gmail.com/
    
    The discussion centers on an issue with mdadm where attempting to
    assemble a RAID array caused a null pointer dereference. The problem
    was traced to a missing update_super() function in super-ddf.c, which
    led to a crash in Assemble.c.
    
    Signed-off-by: lilinzhe <llz@antiy.cn>
```

slayercat commented on super-ddf.c:
```diff
@@ -5195,6 +5195,19 @@ static void default_geometry_ddf(struct supertype *st, int *level, int *layout,
+	 * A dummy update_super function is needed
+	 * to enable the mounting feature.
+	 */
+	pr_err("%s: not implemented\n", __func__);
```
Thank you for addressing that. I have converted this to use dprintf.

slayercat commented:
Thank you very much for your prompt response and detailed guidance. I have addressed your comments and corrected the submission accordingly.

If there is anything I may have misunderstood or any other areas that need further adjustments, please let me know.

mtkaczyk commented:
Thanks @slayercat,
I'm sick now and Christmas time is comming. I will proceed with it after holidays.

slayercat commented:
Take care~

I hope you feel better soon! 

Wishing you a joyful Christmas season filled with warmth and happiness. Rest up and take your time.

Looking forward to catching up after the holidays.

mtkaczyk commented on super-ddf.c:
```diff
@@ -877,30 +881,178 @@ static void *load_section(int fd, struct ddf_super *super, void *buf,
+		}
+
+		/* Search for the magic value within the read block */
+		for (size_t offset = 0;
```
Please note that tab size should be `8`

mtkaczyk commented on super-ddf.c:
```diff
@@ -877,30 +881,178 @@ static void *load_section(int fd, struct ddf_super *super, void *buf,
+				  unsigned long long *out)
 {
 	unsigned long long dsize;
+	unsigned long long search_start;
```
variables should be declared in reverse Christmas tree convention:
```c
	unsigned long long search_start;
	unsigned long long search_end;
	size_t bytes_block_to_read;
	unsigned long long dsize;
	unsigned long long pos;
	int bytes_current_read;
	be32 *magic_ptr = NULL;
	void *buffer = NULL;

	int result = 0;
```
This applied to the all change. It is not critical but I have to point you that.

mtkaczyk reviewed (commented):
LGTM, few small comments. I run review to see if tests are failing in our CI.

mtkaczyk commented:
@slayercat I don't understand why but with your change tests execution looks like blocked. I looked into action tab and our time scheduled execution is performing correctly, for example:
https://github.com/md-raid-utilities/mdadm/actions/runs/12738826719

I cannot take it until it is solved. You need to debug test script if you want to merge it, sorry :(
In CI nothing is logged. It may mean that first test is blocked `00confnames` (test result is logged after test execution).

slayercat commented:
Thank you for your response.

Alright, I will take some time to confirm the execution details of the script. I truly appreciate your reply.

slayercat commented:
Thank you very much for your reminder. I have addressed the issue that was causing an infinite loop, which I had previously overlooked.

I have completed all the unit tests on a CentOS virtual machine locally. Some of the tests seem to fail randomly. I have attached the output and relevant logs from /var/tmp for your reference.

The upcoming period is a public holiday. If you have any questions, please leave a message, and I will respond as promptly as possible.

Thank you again, and I wish you happiness and prosperity.

[unittest_log.tgz](https://github.com/user-attachments/files/18521126/unittest_log.tgz)

mtkaczyk reviewed (approved):
@slayercat great contribution!

mtkaczyk commented:
No more feedback. This change is limited to DDF (no interest from others).
Merging.
